app/src/common/shared/org/mozilla/vrbrowser/History.java

```diff
@@ -0,0 +1,210 @@
+            title = aObject.optString("n", "");
+            timestamp = aObject.optLong("t", System.currentTimeMillis());
+        }
+
+        JSONObject toJSON() {
+            JSONObject obj = new JSONObject();
+            try {
+                obj.put("u", url);
+                obj.put("n", title);
+                obj.put("t", timestamp);
+            }
+            catch (JSONException ex) {
+            }
+            return obj;
+        }
+    }
+
+    public History(Context aContext) {
+        mContext = aContext;
+        mItems = new HashMap<>();
+        SessionStore.get().addProgressListener(this);
+        restore();
+    }
+
+    public void updateContext(Context aContext) {
+        mContext = aContext;
+    }
+
+    public void release() {
+        SessionStore.get().removeProgressListener(this);
+    }
+
+    public String encodedJSON() {
+        String value = getJSONString();
+        String base64 = Base64.encodeToString(value.getBytes(), 0);
+        return Uri.encode(base64);
+    }
+
+    public void clear() {
+        mItems.clear();
+        save();
+    }
+
+    private void save() {
+        JSONArray array = new JSONArray();
+        for (Item item: mItems.values()) {
+            array.put(item.toJSON());
+        }
+
+        try (FileWriter file = new FileWriter(mContext.getFilesDir().getAbsolutePath() + "/" + FILENAME)) {
+            file.write(getJSONString());
+        }
+        catch (Exception ex) {
+            Log.e(LOGTAG, "Error saving history", ex);
+        }
+    }
+
+    private void restore() {
+        try (BufferedReader reader = new BufferedReader(new FileReader(mContext.getFilesDir().getAbsolutePath() + "/" + FILENAME))) {
+            StringBuilder sb = new StringBuilder();
+            String line = reader.readLine();
+
+            while (line != null) {
+                sb.append(line);
+                sb.append("\n");
+                line = reader.readLine();
+            }
+            try {
+                JSONArray array = new JSONArray(sb.toString());
+                mItems.clear();
+                for (int i = 0; i < array.length(); ++i) {
+                    Item item = new Item(array.getJSONObject(i));
+                    if (item.url != null && item.url.length() > 0) {
+                        mItems.put(key(item.url), item);
+                    }
+                }
+
+            } catch (JSONException e) {
+                Log.e(LOGTAG, "Error parsing history", e);
+            }
+        }
+        catch (Exception ex) {
+            Log.e(LOGTAG, "Error reading history", ex);
+        }
+    }
+
+    private String getJSONString() {
+        JSONArray array = new JSONArray();
+        for (Item item: mItems.values()) {
+            array.put(item.toJSON());
+        }
+        return array.toString();
+    }
+
+    private String key(String aUri) {
+        String result = aUri.toLowerCase();
+        int index = result.indexOf('#');
```

Comment:
it'd be nice if we could use some helper/util method for extracting the hash instead of doing fragile string substring matching and slicing.

out of curiosity, does Java have any URI parsing libraries (like [JavaScript's `URL`](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL) or [Node's `URL`](https://nodejs.org/api/url.html))?